### PR TITLE
from dev branch

### PR DESCRIPTION
**Pull request by YegorPopinako**
implemented all requirements from final project tasks

### CODE REVIEW

File: src/main/java/ua/javarush/yegor/animal/Animal.java [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+import ua.javarush.yegor.island.Area;
+
+import java.util.UUID;
+import java.util.concurrent.ThreadLocalRandom;
+
+@EqualsAndHashCode(onlyExplicitlyIncluded = true)
+public abstract class Animal implements Movable {
+
+    @EqualsAndHashCode.Include
+    private final int id;
```
Comment (Anderwerty) [inline src/main/java/ua/javarush/yegor/animal/Animal.java]: better to use String type here

File: src/main/java/ua/javarush/yegor/animal/Animal.java [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+
+    @Getter
+    @Setter
+    protected int healthPoints;
+
+    @Setter
+    private boolean hasMoved;
+
+    Direction[] directions = Direction.values();
```
Comment (Anderwerty) [inline src/main/java/ua/javarush/yegor/animal/Animal.java]: Why do we need this field?

File: src/main/java/ua/javarush/yegor/animal/Animal.java [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+    }
+
+    public abstract void eat(Area area);
+
+    public abstract void reproduce(Area area);
+
+    public boolean isAlive() {
+        return healthPoints > 0;
+    }
+}
```
Comment (Anderwerty) [inline src/main/java/ua/javarush/yegor/animal/Animal.java]: please add a blank line after last brace

File: src/main/java/ua/javarush/yegor/animal/settings/utils/AnimalSettingsUtil.java [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+        if (clazz.isAnnotationPresent(Settings.class)) {
+            String settingsFile = clazz.getAnnotation(Settings.class).settingsFile();
+
+            try (InputStream inputStream = AnimalUnit.class.getClassLoader().getResourceAsStream(settingsFile)) {
+                if (inputStream != null) {
+                    ObjectMapper mapper = new ObjectMapper();
+                    return mapper.readValue(inputStream, AnimalUnit.class);
+                }
+            } catch (Exception e) {
+                e.getCause();
```
Comment (Anderwerty) [inline src/main/java/ua/javarush/yegor/animal/settings/utils/AnimalSettingsUtil.java]: What is the sense of this line?

File: src/main/java/ua/javarush/yegor/direction/Direction.java [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+    RIGHT(1,0), LEFT(-1, 0), UP(0,1), DOWN(0,-1), NONE(0 ,0);
+
+    Direction(int deltaX, int deltaY) {
+        this.deltaX = deltaX;
+        this.deltaY = deltaY;
+    }
+
+    private final int deltaX;
+    private final int deltaY;
+}
```
Comment (Anderwerty) [inline src/main/java/ua/javarush/yegor/direction/Direction.java]: please add a blank line after last brace

File: src/main/java/ua/javarush/yegor/island/Area.java [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+        plants.add(plant);
+    }
+
+    public void removePlant(Plant plant) {
+        plants.remove(plant);
+    }
+
+    public void reproducePlants() {
+        for(int i = 0; i < ThreadLocalRandom.current().nextInt(600, MAX_PLANT_NUMBER); i++) {
+            if(plants.size() < MAX_PLANT_NUMBER) {
```
Comment (Anderwerty) [inline src/main/java/ua/javarush/yegor/island/Area.java:124]: 600 looks like some magic number

File: src/main/java/ua/javarush/yegor/island/Island.java [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+        StringBuilder sb = new StringBuilder();
+        for (Area[] row : areas) {
+            for (Area area : row) {
+                sb.append("[" + area + "]");
+            }
+            sb.append("\n");
+        }
+        return sb.toString();
+    }
+}
```
Comment (Anderwerty) [inline src/main/java/ua/javarush/yegor/island/Island.java]: the same

File: src/main/java/ua/javarush/yegor/island/IslandController.java [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+    }
+
+    private static void animalsCount(Area area, Map<Class<? extends Animal>, Integer> animalCounts) {
+        Map<Class<? extends Animal>, Set<Animal>> animalsCopy = area.getAnimals();
+        for (var entry : animalsCopy.entrySet()) {
+            Class<? extends Animal> animalClass = entry.getKey();
+            animalCounts.put(animalClass, animalCounts.getOrDefault(animalClass, 0) + entry.getValue().size());
+        }
+    }
+}
```
Comment (Anderwerty) [inline src/main/java/ua/javarush/yegor/island/IslandController.java]: the same

**Comment by Anderwerty** [comment]
@YegorPopinako 
Please add a description to the pull request and visualvm to check that threads work in parallel.

**Comment by YegorPopinako** [comment]
Created all requested classes, including:
**Predators**: Wolf, Boa constrictor, Fox, Bear, Eagle
**Herbivores**: Horse, Deer, Rabbit, Mouse, Goat, Sheep, Boar, Buffalo, Duck, Goose
**Plants**

Implemented mandatory part of the task:
- Hierarchy of Animals
- Animal behaviour
- Multithreading
- Statistics about the state of the island at each iteration(in the console)

### Output to the console:
Animals initializing on the island with the statistic logged to the console:
image
Statistic on day 5(animals eating each other and dying from hunger):
image

**Pull request merged**